### DISCUSSION TITLE
What would be required to implement AP mode?

### DISCUSSION THREAD

**Discussion by the-maldridge** [gokrazy/gokrazy#254] [Q&A]
I'm interested in using gokrazy for a project, but I would need to be able to have the rpi zero 2 w act as a self-contained AP.  What all would I need to be looking at to implement this support?  I read through the issues on implementing wifi client support, and it looks like the security supplicant is the primary difficulty I'd run into.  Since router7 exists, I assume I could pull the dhcp implementation from there, so that just leaves configuring the interface itself.

**Comment by stapelberg** [comment]
Yes, “just configuring the interface” is the main obstacle :) 

Be sure to check out https://gokrazy.org/development/non-go/ to evaluate what you need — you can copy over non-Go software piece by piece until you have a working example, then judge the effort and required tasks from there.

**Comment by damdo** [comment]
@the-maldridge interested in this one as well, please do keep us posted on your results :)
Thanks!

**Comment by mpl** [comment]
Yeah, I need that as well. :)
I thought it wouldn't be possible without NetworkManager and nmcli, but I've just found this as a first lead , so maybe there's hope.
https://wiki.archlinux.org/title/Software_access_point
I really hope I don't have to pull in the whole NetworkManager ecosystem with the freeze method.

**Reply by mpl** [reply]
progress update:
with a combination of hostapd ("installed" with the freeze method) + github.com/rtr7/router7/cmd/dhcp4d , and messing around manually with the ip command, I was able to make a working access point (no bridging, as I don't need/want it in my case) on gokrazy.
I still need to automate it all, and more importantly, make it work alongside /user/wifi (so both an access point and a station), but I think it can be all done. :)

**Comment by mpl** [comment]
Hey @stapelberg ,
As far as I know, to use the rpi both as an access point and as a station, one needs to create virtual interfaces, with something like:
```
iw dev wlan0 interface add wlan0_sta type managed addr 12:34:56:78:ab:cd  
iw dev wlan0 interface add wlan0_ap  type managed addr 12:34:56:78:ab:ce
```
Do you know of any alternative to iw that could do the job on gokrazy (i.e. do you think it can be done in Go)? Or do you think I should try "installing" iw with the freeze method?

**Reply by stapelberg** [reply]
Maybe it could be done in Go (see the https://github.com/gokrazy/wifi package and its dependencies), but freezing iw is probably a lot quicker.

**Reply by mpl** [reply]
ACK, thanks.

I now have a fully working /user/ap service on gokrazy btw.
Still in competition with /gokrazy/wifi (I enable one or the other through the config file for now) though, so yeah I'll try to make them work together with iw then.

And speaking of `github.com/gokrazy/wifi` , I was looking at the code today, pulling bits and pieces to do my AP service, and there's one thing I didn't get. It seems to me that we're iterating on all the possible wireless network interfaces (https://github.com/gokrazy/wifi/blob/main/wifi.go#L58) , but on the other hand everything we're doing elsewhere is tied specifically to wlan0 in several places.
So I'm confused, are all these interfaces related to wlan0 itself, or is wlan0 only one of them?

**Reply by stapelberg** [reply]
> It seems to me that we're iterating on all the possible wireless network interfaces

Don’t read too much into that, I don’t think there’s a deep reason behind it. The mdlayher/wifi package just doesn’t have an InterfaceByName convenience function, and we should probably filter the returned interfaces. But, since in practice we only have `wlan0`, it doesn’t (didn’t?) matter.

> So I'm confused, are all these interfaces related to wlan0 itself, or is wlan0 only one of them?

I can’t answer that. I think it depends on your definition of an interface, perhaps even on which part of the kernel you’re interacting with (WiFi vs. non-WiFi interfaces). Some interfaces represent physical cards (wlan0), others might be virtual interfaces on that same device.
